packages/projection-typeorm/src/TypeormStabilityWindowBuffer.ts

```diff
@@ -0,0 +1,204 @@
+      })
+    ]);
+    this.#tip$.next(tip[0]?.data || 'origin');
+    this.#setTail(tail[0]?.data || 'origin');
+  }
+
+  shutdown(): void {
+    this.#tip$.complete();
+    this.#tail$.complete();
+  }
+
+  async #rollForward(evt: RollForwardEvent<Operators.WithNetworkInfo & WithTypeormContext>) {
+    const { eventType, transactionCommitted$, queryRunner, blockEntity, block } = evt;
+    const {
+      header: { blockNo }
+    } = block;
+    const repository = queryRunner.manager.getRepository(BlockDataEntity);
+    if (eventType === ChainSyncEventType.RollForward) {
+      this.#logger.debug('Add block data at height', blockNo);
+      const blockData = repository.create({
+        block: blockEntity,
+        data: block
+      });
+      await Promise.all([repository.insert(blockData), this.#deleteOldBlockData(evt)]);
+      transactionCommitted$.subscribe(() => {
+        this.#tip$.next(block);
+        if (this.#tail === 'origin') {
+          this.#setTail(block);
+        }
+      });
+    }
+  }
+
+  async #rollBackward({
+    transactionCommitted$,
+    queryRunner,
+    block: {
+      header: { slot, blockNo }
+    }
+  }: RollBackwardEvent<WithBlock & WithTypeormContext>) {
+    const repository = queryRunner.manager.getRepository(BlockDataEntity);
+    // No need to delete rolled back block here, as it should cascade when the block entity gets deleted
+    const prevTip = await repository.findOne({
+      order: {
+        block: {
+          slot: 'DESC'
+        }
+      },
+      relations: {
+        block: true
+      },
+      select: blockDataSelect,
+      where: {
+        block: {
+          slot: LessThan(slot)
+        }
+      }
+    });
+    if (prevTip?.data && blockNo !== prevTip?.data.header.blockNo + 1) {
+      throw new Error('Assert: inconsistent PgStabilityWindowBuffer at rollBackward');
+    }
+    transactionCommitted$.subscribe(() => {
+      this.#tip$.next(prevTip?.data || 'origin');
+      if (!prevTip?.data) {
+        this.#setTail('origin');
+      }
+    });
+  }
+
+  async #deleteOldBlockData({
+    genesisParameters: { securityParameter },
+    block: {
+      header: { blockNo }
+    },
+    queryRunner,
+    transactionCommitted$
+  }: RollForwardEvent<Operators.WithNetworkInfo & WithTypeormContext>) {
+    if (blockNo < securityParameter || blockNo % this.#compactEvery !== 0) {
+      return;
+    }
+    const repository = queryRunner.manager.getRepository(BlockDataEntity);
+    const nextTailBlockHeight = blockNo - securityParameter;
+    const [nextTailEntity] = await Promise.all([
+      repository.findOne({
+        relations: {
+          block: true
+        },
+        select: {
+          block: { height: true },
+          data: true as any,
+          id: true
+        },
+        where: {
+          block: {
+            height: nextTailBlockHeight
+          }
+        }
+      }),
+      // Review: this is fragile, may be better to make 2 type-safe(r) queries instead:
```

Comment:
Let's stick with the most performant option. Either drop the comment or move it to a Jira ticket

Reply:
Dropped the comment e9ad0e196f5bc35ad6a85b22520f0b0a10e8cb44